panic encountered when ctrl+c-ing a query

I have a 4 node roachprod cluster with a TPC-C 100 warehouse dataset, loaded with
` bin/workload fixtures load tpcc  --warehouses=100 'postgres://root@35.190.154.175:26257?sslmode=disable'`

I ran the following query:
`(SELECT ol_w_id, ol_d_id, ol_o_id FROM order_line WHERE ol_delivery_d IS NULL ORDER BY ol_w_id, ol_d_id, ol_o_id DESC)                                                         EXCEPT ALL                                                                                                                                                                                                  (SELECT o_w_id, o_d_id, o_id FROM "order"@primary ORDER BY o_w_id, o_d_id, o_id DESC)`

After approximately 10 seconds, I knew I had done something wrong in constructing my query since I wanted a query that finished quickly, so I ctrl+c-ed out. This caused the node to crash with the following log message:

```
E180310 00:22:29.377528 917 sql/conn_executor.go:512  [n1,client=10.142.0.10:60920,user=root] a SQL panic has occurred while executing "(SELECT ol_w_id, ol_d_id, ol_o_id FROM order_line WHERE ol_delivery_d IS NULL ORDER BY ol_w_id, ol_d_id, ol_o_id DESC) EXCEPT ALL (SELECT o_w_id, o_d_id, o_id FROM \"order\"@primary ORDER BY o_w_id, o_d_id, o_id DESC)": can't overwrite err: write tcp 10.142.0.9:26257->10.142.0.10:60920: write: connection reset by peer with err: write tcp 10.142.0.9:26257->10.142.0.10:60920: write: connection reset by peer
E180310 00:22:29.377757 917 util/log/crash_reporting.go:182  [n1,client=10.142.0.10:60920,user=root] a panic has occurred!
E180310 00:22:29.378065 917 util/log/crash_reporting.go:440  [n1,client=10.142.0.10:60920,user=root] Reported as error 8f4f265d29474b49881eea0716a63983
panic: can't overwrite err: write tcp 10.142.0.9:26257->10.142.0.10:60920: write: connection reset by peer with err: write tcp 10.142.0.9:26257->10.142.0.10:60920: write: connection reset by peer [recovered]
        panic: panic while executing 1 statements: (SELECT _, _, _ FROM _ WHERE _ IS _ ORDER BY _, _, _ DESC) EXCEPT ALL (SELECT _, _, _ FROM _@_ ORDER BY _, _, _ DESC); caused by can't overwrite err: write tcp 10.142.0.9:26257->10.142.0.10:60920: write: connection reset by peer with err: write tcp 10.142.0.9:26257->10.142.0.10:60920: write: connection reset by peer

goroutine 917 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc426576000, 0x2661c60, 0xc426a08500, 0x2038c00, 0xc426da0110)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:526 +0x365
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func2(0xc426576000, 0x2661c60, 0xc426a08500)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:487 +0x61
panic(0x2038c00, 0xc426da0110)
        /usr/local/go/src/runtime/panic.go:505 +0x229
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*commandResult).SetError(0xc4265c8c80, 0x2643460, 0xc424f10190)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/command_result.go:187 +0x134
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc426576000, 0x2661c60, 0xc420b2dac0, 0x26645a0, 0xc420b2da40, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:556 +0x853
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc426576000, 0x2661c60, 0xc420b2dac0, 0x26645a0, 0xc420b2da40, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:342 +0x9c7
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc426576000, 0x2661c60, 0xc420b2dac0, 0x26645a0, 0xc420b2da40, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:83 +0x358
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc426576000, 0x2661c60, 0xc426a08500, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:908 +0x1cc8
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc4203d7dc0, 0x2661c60, 0xc426a08500, 0xc426573054, 0x8, 0xc426573009, 0x4, 0x0, 0x0, 0x264b120, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:489 +0xf53
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl.func3(0xc4203d7dc0, 0x2661c60, 0xc426a08500, 0xc420349500, 0x5400, 0x15000, 0xc420a42cd0, 0xc426e52690, 0xc426cd8c50, 0xc426e52680)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:260 +0xfc
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:259 +0xf14
(END)

```

The cockroach binary is the current head of the release-2.0 branch, so this is worrying:
```
Build Tag:    v2.0-beta.20180305-206-g2329ca0
Build Time:   2018/03/09 15:55:07
Distribution: CCL
Platform:     linux amd64
Go Version:   go1.10
C Compiler:   4.2.1 Compatible Clang 3.8.0 (tags/RELEASE_380/final)
Build SHA-1:  2329ca0f1e926bb7dba8abd558bc4631bbddd04c
Build Type:   development
```

Assigned to @jordanlewis for triage.

--- a/pkg/base/test_server_args.go
+++ b/pkg/base/test_server_args.go
@@ -103,6 +103,11 @@ type TestServerArgs struct {
 	// If set, web session authentication will be disabled, even if the server
 	// is running in secure mode.
 	DisableWebSessionAuthentication bool
+
+	// ConnResultsBufferBytes is the size of the buffer in which each connection
+	// accumulates results set. Results are flushed to the network when this
+	// buffer overflows.
+	ConnResultsBufferBytes int
 }
 
 // TestClusterArgs contains the parameters one can set when creating a test

--- a/pkg/server/config.go
+++ b/pkg/server/config.go
@@ -77,7 +77,7 @@ const (
 	minimumNetworkFileDescriptors     = 256
 	recommendedNetworkFileDescriptors = 5000
 
-	defaultConnResultsBufferBytes = 16 << 10
+	defaultConnResultsBufferBytes = 16 << 10 // 16 KiB
 )
 
 var productionSettingsWebpage = fmt.Sprintf(

--- a/pkg/server/config.go
+++ b/pkg/server/config.go
@@ -76,6 +76,8 @@ const (
 
 	minimumNetworkFileDescriptors     = 256
 	recommendedNetworkFileDescriptors = 5000
+
+	defaultConnResultsBufferBytes = 16 << 10
 )
 
 var productionSettingsWebpage = fmt.Sprintf(
@@ -241,6 +243,11 @@ type Config struct {
 	// deleted.
 	UseLegacyConnHandling bool
 
+	// ConnResultsBufferBytes is the size of the buffer in which each connection
+	// accumulates results set. Results are flushed to the network when this
+	// buffer overflows.
+	ConnResultsBufferBytes int
+
 	enginesCreated bool
 }
 
@@ -378,6 +385,7 @@ func MakeConfig(ctx context.Context, st *cluster.Settings) Config {
 		},
 		TempStorageConfig: base.TempStorageConfigFromEnv(
 			ctx, st, storeSpec, "" /* parentDir */, base.DefaultTempStorageMaxSizeBytes),
+		ConnResultsBufferBytes: defaultConnResultsBufferBytes,
 	}
 	cfg.AmbientCtx.Tracer = st.Tracer
 

--- a/pkg/server/server.go
+++ b/pkg/server/server.go
@@ -516,9 +516,11 @@ func NewServer(cfg Config, stopper *stop.Stopper) (*Server, error) {
 			s.distSender,
 			s.gossip,
 			s.stopper,
-			sqlExecutorTestingKnobs.DistSQLPlannerKnobs),
-		ExecLogger:  log.NewSecondaryLogger(nil, "sql-exec", true /*enableGc*/, false /*forceSyncWrites*/),
-		AuditLogger: log.NewSecondaryLogger(s.cfg.SQLAuditLogDirName, "sql-audit", true /*enableGc*/, true /*forceSyncWrites*/),
+			sqlExecutorTestingKnobs.DistSQLPlannerKnobs,
+		),
+		ExecLogger:             log.NewSecondaryLogger(nil, "sql-exec", true /*enableGc*/, false /*forceSyncWrites*/),
+		AuditLogger:            log.NewSecondaryLogger(s.cfg.SQLAuditLogDirName, "sql-audit", true /*enableGc*/, true /*forceSyncWrites*/),
+		ConnResultsBufferBytes: s.cfg.ConnResultsBufferBytes,
 	}
 
 	if sqlSchemaChangerTestingKnobs := s.cfg.TestingKnobs.SQLSchemaChanger; sqlSchemaChangerTestingKnobs != nil {

--- a/pkg/server/testserver.go
+++ b/pkg/server/testserver.go
@@ -203,6 +203,10 @@ func makeTestConfigFromParams(params base.TestServerArgs) Config {
 	}
 	cfg.TestingKnobs.Store.(*storage.StoreTestingKnobs).SkipMinSizeCheck = true
 
+	if params.ConnResultsBufferBytes != 0 {
+		cfg.ConnResultsBufferBytes = params.ConnResultsBufferBytes
+	}
+
 	return cfg
 }
 

--- a/pkg/sql/conn_executor_exec.go
+++ b/pkg/sql/conn_executor_exec.go
@@ -581,7 +581,6 @@ func (ex *connExecutor) dispatchToExecutionEngine(
 	}
 	planner.statsCollector.PhaseTimes()[plannerEndExecStmt] = timeutil.Now()
 	if err != nil {
-		res.SetError(err)
 		return err
 	}
 	recordStatementSummary(
@@ -598,6 +597,7 @@ func (ex *connExecutor) dispatchToExecutionEngine(
 // execWithLocalEngine runs a plan using the local (non-distributed) SQL
 // engine.
 // If an error is returned, the connection needs to stop processing queries.
+// Such errors are also written to res.
 // Query execution errors are written to res; they are not returned.
 func (ex *connExecutor) execWithLocalEngine(
 	ctx context.Context, planner *planner, stmtType tree.StatementType, res RestrictedCommandResult,
@@ -639,6 +639,7 @@ func (ex *connExecutor) execWithLocalEngine(
 			return commErr
 		})
 		if commErr != nil {
+			res.SetError(commErr)
 			return commErr
 		}
 		if queryErr != nil {

--- a/pkg/sql/executor.go
+++ b/pkg/sql/executor.go
@@ -279,6 +279,11 @@ type ExecutorConfig struct {
 	// Caches updated by DistSQL.
 	RangeDescriptorCache *kv.RangeDescriptorCache
 	LeaseHolderCache     *kv.LeaseHolderCache
+
+	// ConnResultsBufferBytes is the size of the buffer in which each connection
+	// accumulates results set. Results are flushed to the network when this
+	// buffer overflows.
+	ConnResultsBufferBytes int
 }
 
 // Organization returns the value of cluster.organization.

--- a/pkg/sql/pgwire/conn.go
+++ b/pkg/sql/pgwire/conn.go
@@ -1047,7 +1047,7 @@ func (c *conn) Flush(pos sql.CmdPos) error {
 // maybeFlush flushes the buffer to the network connection if it exceeded
 // connResultsBufferSizeBytes.
 func (c *conn) maybeFlush(pos sql.CmdPos) (bool, error) {
-	if c.writerState.buf.Len() <= connResultsBufferSizeBytes {
+	if c.writerState.buf.Len() <= c.execCfg.ConnResultsBufferBytes {
 		return false, nil
 	}
 	return true, c.Flush(pos)

--- a/pkg/sql/pgwire/conn_test.go
+++ b/pkg/sql/pgwire/conn_test.go
@@ -36,6 +36,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlutil"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/util"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
@@ -615,3 +616,72 @@ func (l pgxTestLogger) Log(level pgx.LogLevel, msg string, data map[string]inter
 
 // pgxTestLogger implements pgx.Logger.
 var _ pgx.Logger = pgxTestLogger{}
+
+// Test that closing a client connection such that producing results rows
+// encounters network errors doesn't crash the server (#23694).
+//
+// We'll run a query that produces a bunch of rows and close the connection as
+// soon as the client received anything, this way ensuring that:
+// a) the query started executing when the connection is closed, and so it's
+// likely to observe a network error and not a context cancelation, and
+// b) the connection's server-side results buffer has overflowed, and so
+// attempting to produce results (through CommandResult.AddRow()) observes
+// network errors.
+func TestConnClose(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	s, db, _ := serverutils.StartServer(t, base.TestServerArgs{
+		// Make the connections' results buffers really small so that it overflows
+		// when we produce a few results.
+		ConnResultsBufferBytes: 10,
+		// Andrei is too lazy to figure out the incantation for telling pgx about
+		// our test certs.
+		Insecure: true,
+	})
+	ctx := context.TODO()
+	defer s.Stopper().Stop(ctx)
+
+	r := sqlutils.MakeSQLRunner(db)
+	r.Exec(t, "CREATE DATABASE test")
+	r.Exec(t, "CREATE TABLE test.test AS SELECT * FROM generate_series(1,100)")
+
+	host, ports, err := net.SplitHostPort(s.ServingAddr())
+	if err != nil {
+		t.Fatal(err)
+	}
+	port, err := strconv.Atoi(ports)
+	if err != nil {
+		t.Fatal(err)
+	}
+	// We test both with and without DistSQL, as the way that network errors are
+	// observed depends on the engine.
+	testutils.RunTrueAndFalse(t, "useDistSQL", func(t *testing.T, useDistSQL bool) {
+		conn, err := pgx.Connect(pgx.ConnConfig{
+			Host:     host,
+			Port:     uint16(port),
+			User:     "root",
+			Database: "system",
+		})
+		if err != nil {
+			t.Fatal(err)
+		}
+		var query string
+		if useDistSQL {
+			query = `SET DISTSQL = 'always'`
+		} else {
+			query = `SET DISTSQL = 'off'`
+		}
+		if _, err := conn.Exec(query); err != nil {
+			t.Fatal(err)
+		}
+		rows, err := conn.Query("SELECT * FROM test.test")
+		if err != nil {
+			t.Fatal(err)
+		}
+		if hasResults := rows.Next(); !hasResults {
+			t.Fatal("expected results")
+		}
+		if err := conn.Close(); err != nil {
+			t.Fatal(err)
+		}
+	})
+}

--- a/pkg/sql/pgwire/v3.go
+++ b/pkg/sql/pgwire/v3.go
@@ -48,10 +48,6 @@ const (
 	authCleartextPassword int32 = 3
 )
 
-// connResultsBufferSizeBytes refers to the size of the result set which we
-// buffer into memory prior to flushing to the client.
-const connResultsBufferSizeBytes = 16 << 10
-
 // readTimeoutConn overloads net.Conn.Read by periodically calling
 // checkExitConds() and aborting the read if an error is returned.
 type readTimeoutConn struct {
@@ -1266,7 +1262,7 @@ func (c *v3Conn) flush(forceSend bool) error {
 		return nil
 	}
 
-	if forceSend || state.buf.Len() > connResultsBufferSizeBytes {
+	if forceSend || state.buf.Len() > c.executor.Cfg().ConnResultsBufferBytes {
 		state.hasSentResults = true
 		state.txnStartIdx = 0
 		if _, err := state.buf.WriteTo(c.wr); err != nil {

